test/test_swallow.py

```diff
@@ -0,0 +1,141 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from libqtile import config
+from libqtile.backend.x11 import xcbq
+from libqtile.backend.x11.core import Core
+from libqtile.confreader import Config
+from libqtile.lazy import lazy
+
+
+# Function that does nothing
+def swallow_nop(qtile):
+    pass
+
+
+# Config with multiple keys and swallow parameters
+class SwallowConfig(Config):
+    keys = [
+        config.Key(
+            ["control"],
+            "k",
+            lazy.function(swallow_nop),
+        ),
+        config.Key(
+            ["control"],
+            "j",
+            lazy.function(swallow_nop),
+            swallow=False
+        ),
+        config.Key(
+            ["control"],
+            "i",
+            lazy.function(swallow_nop).when(layout='idonotexist')
+        ),
+        config.Key(
+            ["control"],
+            "o",
+            lazy.function(swallow_nop).when(layout='idonotexist'),
+            lazy.function(swallow_nop)
+        ),
+    ]
+
+    mouse = [
+        config.Click(
+            [],
+            "Button1",
+            lazy.function(swallow_nop),
+        ),
+        config.Click(
+            ["control"],
+            "Button3",
+            lazy.function(swallow_nop),
+            swallow=False
+        ),
+        config.Click(
+            ["mod4"],
+            "Button3",
+            lazy.function(swallow_nop).when(layout='idonotexist')
+        ),
+        config.Click(
+            [],
+            "Button3",
+            lazy.function(swallow_nop).when(layout='idonotexist'),
+            lazy.function(swallow_nop)
+        ),
+    ]
+
+
+# Helper to send process_key_event to the core manager
+# It also looks up the keysym and mask to pass to it
+def send_process_key_event(manager, key):
+    keysym, mask = Core.lookup_key(None, key)
+    output = manager.c.eval(f"self.process_key_event({keysym}, {mask})")
+    # Assert if eval successful
+    assert output[0]
+    # Convert the string to a bool
+    return output[1] == 'True'
+
```

Comment:
I wonder if we could replace this with `cmd_simulate_buttonpress` (once the wayland implementation is added)

Reply:
Probably indeed! The way I did it here is really hacky

Reply:
Ha, we have lots of these hacky evals in the tests - they are very useful!

Reply:
True but it's especially hacky because I used None for the self argument in lookup key :laughing:

Reply:
On second thought, the `cmd_simulate_keypress` function does not return the handled handled boolean which is needed here. So that will have to be added there

Reply:
The test window gets closed if it receives a key event, so if it's swallow vs non-swallow that you need to check then that might be a useful way to detect if a key event was not swallowed.

Reply:
TIL that is useful!